api/src/zenbox_smt/mod.rs

```diff
@@ -0,0 +1,226 @@
+use crate::ginger_calls::{FieldElement, GingerSMT, GingerMHTPath, Error, get_ginger_smt, flush_ginger_smt, set_ginger_smt_persistency, get_position_in_ginger_smt, is_position_empty_in_ginger_smt, add_leaf_to_ginger_smt, remove_leaf_from_ginger_smt, get_ginger_smt_root, get_ginger_smt_path, get_leaf_from_ginger_smt};
+use crate::zenbox_smt::bitvector_smt::*;
+use algebra::UniformRand;
+use rand::rngs::OsRng;
+
+pub mod bitvector_smt;
+
+pub struct ZenBoxSMT{
+    state_smt: GingerSMT,
+    bitvector_smt: BitvectorSMT,
+    base_path: String // base path to state_smt and bitvector_smt DBs directories
+}
+
+const BVT_PATH_SUFFIX: &str = "_bvt";
+
+pub fn get_zenbox_smt(state_height: usize, db_path: &str) -> Result<ZenBoxSMT, Error> {
+    assert!(state_height > BVT_LEAF_SIZE_LOG2);
+    // Bitvector needs BVT_LEAF_SIZE times lesser capacity of MT than State
+    let bvt_height = state_height - BVT_LEAF_SIZE_LOG2;
+    Ok(
+        ZenBoxSMT{
+            state_smt: get_ginger_smt(state_height, db_path)?,
+            bitvector_smt: get_bvt(bvt_height, &(db_path.to_owned() + BVT_PATH_SUFFIX))?,
+            base_path: db_path.to_owned()
+        }
+    )
+}
+
+// Just removing existing BVT and creating the new empty one at the same filesystem location
+pub fn reset_bitvector(mut tree: ZenBoxSMT) -> Result<ZenBoxSMT, Error> {
+    let bvt_height = tree.bitvector_smt.height();
+    // Deleting existing BVT DB
+    set_bvt_persistency(&mut tree.bitvector_smt, false);
+    drop(tree.bitvector_smt); // to trigger deletion of BVT-containing directories
+    Ok(
+        ZenBoxSMT{
+            state_smt: tree.state_smt,
+            bitvector_smt: get_bvt(bvt_height, &(tree.base_path.to_owned() + BVT_PATH_SUFFIX))?,
+            base_path: tree.base_path
+        }
+    )
+}
+
+pub fn flush_zenbox_smt(tree: &mut ZenBoxSMT) {
+    flush_ginger_smt(&mut tree.state_smt);
+    flush_bvt(&mut tree.bitvector_smt)
+}
+
+pub fn set_zenbox_smt_persistency(tree: &mut ZenBoxSMT, persistency: bool) {
+    set_ginger_smt_persistency(&mut tree.state_smt, persistency);
+    set_bvt_persistency(&mut tree.bitvector_smt, persistency);
+}
+
+pub fn get_position_in_zenbox_smt(tree: &ZenBoxSMT, leaf: &FieldElement) -> u64 {
+    get_position_in_ginger_smt(&tree.state_smt, leaf)
+}
+
+pub fn is_position_empty_in_zenbox_smt(tree: &ZenBoxSMT, position: u64) -> bool {
+    is_position_empty_in_ginger_smt(&tree.state_smt, position)
+}
+
+pub fn is_box_spent(tree: &ZenBoxSMT, position: u64) -> bool {
+    get_bvt_bit(&tree.bitvector_smt, position)
+}
+
+pub fn add_box_to_zenbox_smt(tree: &mut ZenBoxSMT, box_: &FieldElement, position: u64){
+    // avoid overwriting already existing boxes
+    if is_position_empty_in_ginger_smt(&tree.state_smt, position){
+        // if corresponding bit in BVT is set, this means that current box has already been spent
+        assert!(!get_bvt_bit(&tree.bitvector_smt, position));
+        add_leaf_to_ginger_smt(&mut tree.state_smt, box_, position)
```

Comment:
Shouldn't we set the corresponding bit in the bvt also in the case of new created boxes ?

Reply:
As I understand the BVT functionality, it should track the Boxes that were spent - 1 or never spent - 0. But I will clarify this.